File: src/lib/core/vault/obsidian.ts
import { promises as fs } from "node:fs";
import path from "node:path";
import { parseFrontmatter, serializeFrontmatter } from "./frontmatter";
import type { Vault, VaultNote, VaultSearchHit, VaultWriteOptions } from "./types";

/**
 * Obsidian-compatible vault backed by a folder on disk. Notes are plain `.md`
 * files with optional frontmatter, so the entire agent memory is browsable and
 * editable in Obsidian (point an Obsidian vault at TIRUNO_VAULT_DIR).
 */
export class ObsidianVault implements Vault {
  readonly name = "obsidian";
  private readonly root: string;

  constructor(root?: string) {
    this.root = path.resolve(root || process.env.TIRUNO_VAULT_DIR || path.join(process.cwd(), "vault"));
  }

  /** Resolve + sandbox a vault-relative path to an absolute path. */
  private abs(rel: string): string {
    const clean = rel.replace(/^[/\\]+/, "");
    const full = path.resolve(this.root, clean.endsWith(".md") ? clean : `${clean}.md`);
    if (!full.startsWith(this.root)) throw new Error(`vault path escapes root: ${rel}`);
    return full;
  }

  private rel(abs: string): string {
    return path.relative(this.root, abs).split(path.sep).join("/");
  }

  async read(p: string): Promise<VaultNote | null> {
    try {
      const abs = this.abs(p);
      const [raw, stat] = await Promise.all([fs.readFile(abs, "utf8"), fs.stat(abs)]);
      const { frontmatter, content } = parseFrontmatter(raw);
      return { path: this.rel(abs), frontmatter, content, updatedAt: stat.mtimeMs };
    } catch {
      return null;
    }
  }

  async write(p: string, content: string, opts: VaultWriteOptions = {}): Promise<void> {
    const abs = this.abs(p);
    await fs.mkdir(path.dirname(abs), { recursive: true });
    await fs.writeFile(abs, serializeFrontmatter(opts.frontmatter ?? {}, content), "utf8");
  }

  async append(p: string, content: string): Promise<void> {
    const existing = await this.read(p);
    if (existing) {
      const merged = `${existing.content.replace(/\s+$/, "")}\n\n${content}`;
      await this.write(p, merged, { frontmatter: existing.frontmatter });
    } else {
      await this.write(p, content);
    }
  }

  async list(dir = ""): Promise<string[]> {
    const base = path.resolve(this.root, dir.replace(/^[/\\]+/, ""));
    const out: string[] = [];
    const walk = async (d: string): Promise<void> => {
      let entries: import("node:fs").Dirent[];
      try {
        entries = await fs.readdir(d, { withFileTypes: true });
      } catch {
        return;
      }
      for (const e of entries) {
        const full = path.join(d, e.name);
        if (e.isDirectory()) await walk(full);
        else if (e.isFile() && e.name.endsWith(".md")) out.push(this.rel(full));
      }
    };
    await walk(base);
    return out.sort();
  }

  async search(query: string, opts: { dir?: string; limit?: number } = {}): Promise<VaultSearchHit[]> {
    const terms = query.toLowerCase().split(/\s+/).filter(Boolean);
    if (!terms.length) return [];
    const paths = await this.list(opts.dir ?? "");
    const hits: VaultSearchHit[] = [];

    for (const p of paths) {
      const note = await this.read(p);
      if (!note) continue;
      const hay = `${JSON.stringify(note.frontmatter)} ${note.content}`.toLowerCase();
      let score = 0;
      for (const t of terms) {
        const matches = hay.split(t).length - 1;
        score += matches;
      }
      if (score > 0) {
        const idx = note.content.toLowerCase().indexOf(terms[0]);
        const start = Math.max(0, idx - 60);
        const snippet = note.content.slice(start, start + 200).replace(/\s+/g, " ").trim();
        hits.push({ path: p, score, snippet, frontmatter: note.frontmatter });
      }
    }
    hits.sort((a, b) => b.score - a.score);
    return hits.slice(0, opts.limit ?? 10);
  }

  async remove(p: string): Promise<void> {
    try {
      await fs.unlink(this.abs(p));
    } catch {
      /* already gone */
    }
  }
}

File: worker_py/vault/obsidian.py
"""Obsidian-compatible vault backed by a folder on disk.
Port of ``src/lib/core/vault/obsidian.ts``.

Notes are plain ``.md`` files with optional frontmatter, browsable/editable in
Obsidian (point a vault at ``TIRUNO_VAULT_DIR``).
"""

import json
import os
import re
from dataclasses import dataclass, field
from typing import Any, Dict, List, Optional

from .frontmatter import parse_frontmatter, serialize_frontmatter


@dataclass
class VaultNote:
    path: str
    frontmatter: Dict[str, Any]
    content: str
    updated_at: float


@dataclass
class VaultSearchHit:
    path: str
    score: int
    snippet: str
    frontmatter: Dict[str, Any] = field(default_factory=dict)


class ObsidianVault:
    name = "obsidian"

    def __init__(self, root: Optional[str] = None) -> None:
        self.root = os.path.abspath(
            root or os.environ.get("TIRUNO_VAULT_DIR") or os.path.join(os.getcwd(), "vault")
        )

    def _abs(self, rel: str) -> str:
        clean = re.sub(r"^[/\\]+", "", rel)
        name = clean if clean.endswith(".md") else f"{clean}.md"
        full = os.path.abspath(os.path.join(self.root, name))
        if not full.startswith(self.root):
            raise ValueError(f"vault path escapes root: {rel}")
        return full

    def _rel(self, abs_path: str) -> str:
        return os.path.relpath(abs_path, self.root).replace(os.sep, "/")

    async def read(self, p: str) -> Optional[VaultNote]:
        try:
            full = self._abs(p)
            with open(full, "r", encoding="utf-8") as fh:
                raw = fh.read()
            st = os.stat(full)
            frontmatter, content = parse_frontmatter(raw)
            return VaultNote(
                path=self._rel(full),
                frontmatter=frontmatter,
                content=content,
                updated_at=st.st_mtime * 1000,
            )
        except Exception:  # noqa: BLE001
            return None

    async def write(self, p: str, content: str, frontmatter: Optional[Dict[str, Any]] = None) -> None:
        full = self._abs(p)
        os.makedirs(os.path.dirname(full), exist_ok=True)
        with open(full, "w", encoding="utf-8") as fh:
            fh.write(serialize_frontmatter(frontmatter or {}, content))

    async def append(self, p: str, content: str) -> None:
        existing = await self.read(p)
        if existing:
            merged = re.sub(r"\s+$", "", existing.content) + "\n\n" + content
            await self.write(p, merged, existing.frontmatter)
        else:
            await self.write(p, content)

    async def list(self, dir: str = "") -> List[str]:
        base = os.path.abspath(os.path.join(self.root, re.sub(r"^[/\\]+", "", dir)))
        out: List[str] = []
        for root, _dirs, files in os.walk(base):
            for fname in files:
                if fname.endswith(".md"):
                    out.append(self._rel(os.path.join(root, fname)))
        out.sort()
        return out

    async def search(self, query: str, dir: str = "", limit: int = 10) -> List[VaultSearchHit]:
        terms = [t for t in query.lower().split() if t]
        if not terms:
            return []
        paths = await self.list(dir)
        hits: List[VaultSearchHit] = []

        for p in paths:
            note = await self.read(p)
            if not note:
                continue
            hay = f"{json.dumps(note.frontmatter)} {note.content}".lower()
            score = sum(hay.count(t) for t in terms)
            if score > 0:
                idx = note.content.lower().find(terms[0])
                start = max(0, idx - 60)
                snippet = re.sub(r"\s+", " ", note.content[start : start + 200]).strip()
                hits.append(VaultSearchHit(path=p, score=score, snippet=snippet, frontmatter=note.frontmatter))

        hits.sort(key=lambda h: h.score, reverse=True)
        return hits[:limit]

    async def remove(self, p: str) -> None:
        try:
            os.unlink(self._abs(p))
        except FileNotFoundError:
            pass
